File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0],b[0]));//sort wrt a,b
        List<int[]> result = new ArrayList<>();
        int prev[]=intervals[0];//[1,3]
        for(int i=1;i<intervals.length;i++){
            int []cur=intervals[i];//[2,6]
            if(cur[0]<=prev[1]){//[2]<=[3]
                prev[1]=Math.max(prev[1],cur[1]);//[3] or [6]
            }
            else{
                result.add(prev);
                prev=cur;
            }
        }
        result.add(prev);
        return result.toArray(new int[result.size()][]);
    }
}

File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(),intervals.end(),[](const vector<int> &a,const vector<int> &b){
           return a[0]<b[0]; 
        });
        vector<vector<int>> res;
        vector<int> prev=intervals[0];
        for(int i=1;i<intervals.size();i++){
            vector<int> cur=intervals[i];
            if(prev[1]>=cur[0]){
                prev[1]=max(prev[1],cur[1]);
            }
            else{
                res.push_back(prev);
                prev=cur;
            }
        }
        res.push_back(prev);
        return res;
    }
};